### ISSUE TITLE
hard bouncing emails not added to suppression list?

### ISSUE THREAD

**Issue by tchwpkgorg** [atech/postal#786]
We had the user hard fail 4 times and still over past 4 days:

![Screenshot_20190405_180026](https://user-images.githubusercontent.com/1365984/55616793-974f9c80-57cd-11e9-8d18-69ffd95199d8.png)

And still, user email address is not on suppression list:

```
MariaDB [postal-server-1]> select * from  suppressions WHERE address = 'someone-bouncing@email.tld';
Empty set (0.00 sec)
```

**Comment by willpower232** [comment]
Are you able to share the reason they hard failed with a screenshot from the individual messages page?

**Comment by tchwpkgorg** [comment]
User does not exist:

Permanent SMTP delivery error when sending to aspmx.l.google.com (2a00:1450:4013:c00::1b) Recipient added to suppression list (too many hard fails).

```
550-5.1.1 The email account that you tried to reach does not exist. Please try 550-5.1.1 double-checking the recipient's email address for typos or 550-5.1.1 unnecessary spaces. Learn more at 550 5.1.1 https://support.google.com/mail/?p=NoSuchUser k8si1534884ejk.54 - gsmtp

Time: 0.05s

Support Ref: INLPCORJ
```

**Comment by tchwpkgorg** [comment]
550 No Such User Here:

![bounce](https://user-images.githubusercontent.com/1365984/55637955-17dbc080-5801-11e9-82a0-cac13462c295.png)

**Comment by tchwpkgorg** [comment]
Some users with 4+ "hard fail" emails are added to suppression list. Some are not.

Interestingly, even if it's the same server (i.e. @gmail.com), same bounce cause - some users are added to suppression list, and some are not.

**Comment by tchwpkgorg** [comment]
As I've checked, bouncing emails are sometimes put on hold (i.e. email on suppression list) as advertised, after 2 bounces. Sometimes - after 5 bounces. Sometimes a recipient with 10 bounces keeps getting mail sent from postal.

This is highly irregular behaviour.

**Comment by catphish** [comment]
Your messages are just over 24 hours apart. Addresses are only suppressed for multiple hard failures within 24 hours.

**Comment by tchwpkgorg** [comment]
Is it configurable anywhere?

I'd say most websites don't try to send an email to users more often than daily. If Postal only suppresses emails bouncing multiple times a day - how can we suppress those which got a bounce on a daily/weekly/monthly user notification (assuming there was no successful in the meantime)?

**Comment by willpower232** [comment]
I think the responsible line of code is here https://github.com/atech/postal/blob/master/app/jobs/unqueue_message_job.rb#L395

Arguably you wouldn't want to add people to the suppression list unnecessarily if their email provider was broken temporarily but at the same time those sending daily emails cannot guarantee that their emails will be sent exactly 24 hours apart (with precision down to the second) so this should probably be easier to configure than changing the line of code. Its not an easy decision either way.

**Comment by catphish** [comment]
The answer to this question lies in the reason this feature exists in the first place, and the fact that Postal is designed for transactional emails, not regular ones.

There is really no harm if a daily weekly email bounces every time. What this feature aims to prevent is mailservers going mad and sending a large volume of email to a recipient that is actively trying to reject it.

With that said, the suppression list is a very useful feature, and one that I feel should definitely be expanded to allow:

1) Manual addition and removal of addresses
2) Configurable automated addition. Making the number of failures and timescale configurable would actulally be quite easy.

**Comment by tchwpkgorg** [comment]
> There is really no harm if a daily weekly email bounces every time

There really is. Big email providers (gmail, yahoo, hotmail etc.) don't like email sent to accounts which no longer exist. If you send mail to dead addresses - you will be treated as one which sends spam (depends on ratios and statistics and other algorithms).


Our usage case here is:
- we have a website with tens of thousands of users and send a daily/weekly/monthly email which summarizes some account actions
- we are/were using a different "sending engine" (SparkPost - it will suppress the bouncing recipients)
- we now send a subset of messages through Postal to see how it works

Result: email for big providers go to spambox, although we have all recommended policies in place. Postal dashboard shows ~10% bounces (our website works for over 10 years, so it's natural to have some accounts dead). We keep repeating sending messages which will bounce.


Why the "addresses are only suppressed for multiple hard failures within 24 hours"? It's counterproductive. I understand the point about occasional email bouncing (MX down, user over quota etc.). It is more likely that an email issue at the destination, if it really exists, is fixed when more hours pass, not less.

IMO postal should not look at "suppress only if 2 messages bounced in 24 hour period". It should just keep a bounce count, possibly configurable, default to 2, and it shouldn't matter if the message bounced over 24 hours, 1 week or 1 month period.

To sum up and make it super simple:

- any bouncing message would increase the bounce counter for a given email
- any successful delivery would reset the bounce counter to zero, for a given email
- if bounce counter is 2 (configurable) - suspend delivery to this address
